### PR TITLE
Do not expand theory definitions at the beginning of preprocessing

### PR DESCRIPTION
**Pull request by ajreynol**
This updates the preprocessor so that expand definitions does not expand theory symbols at the beginning of preprocessing.

This also restores the previous expandDefinitions method in arithmetic, which is required for correctly interpreting division by zero in models, but should not be applied at the beginning of preprocessing.  Moreover it ensures that only partial operators are eliminated in arithmetic expandDefinitions, which required an additional argument `partialOnly` to arith::OperatorElim.

This adds `-q` to suppress warnings for many quantified regressions which now emit warnings with `--check-model`.  This will be addressed later as part of https://github.com/CVC4/cvc4-wishues/issues/43.

The purpose of this PR is two-fold:
(1) Currently our responses to `get-value` are incorrect for partial operators like `div`, `mod`, `seq.nth` since partial operators can be left unevaluated.
(2) The preprocessor should have the opportunity to rewrite and eliminate extended operators before they are expanded. This is required for addressing performance issues for non-linear arithmetic.  It is also required for ensuring that trigger selection can be done properly for datatype selectors (to be addressed on a later PR).

### CODE REVIEW

File: src/theory/bv/theory_bv.cpp
Comment (yoni206) [inline src/theory/bv/theory_bv.cpp:203]: Does this change the default behavior for the BV theory? That is, when will `texp` be null?

File: src/theory/fp/theory_fp.cpp
Comment (yoni206) [inline src/theory/fp/theory_fp.cpp:456]: ditto.

File: src/theory/bv/theory_bv.cpp
Comment (ajreynol) [reply src/theory/bv/theory_bv.cpp:203]: It impacts when expand definitions is applied during preprocessing.  It was previously the case that expandDefinitions (for BV) is applied at the beginning of preprocessing during expandDefinitions, now it is applied at the end (during theory-preprocess / ppRewrite).

**Review by yoni206 — COMMENTED** [review]
I am really looking forward to this getting merged :) I think it will really enable better div/mod solving, and in general simpler pps.

Two comments:
1. For BV does it change the default behavior (see comments inline)? If so, perhaps it is worth to verify with @aniemetz and @mpreiner before merging?
2. I am not sure about adding `-q` to that many tests. Does it make sense to first add the PR that will fix that, and only then merge this PR without the `-q`s?

**Comment by ajreynol** [comment]
For (1), to summarize: viewing the preprocessor as a black box, this does not impact BV (or other theories).  

For (2), Fixing the `-q` requires significant changes to the way `--check-models` works.  The order of this PR is intentional.  It is better to fix the preprocessor first, otherwise the design of `--check-models` is a moving target (since the preprocessor currently is not ideal).

**Comment by yoni206** [comment]
(2) -- got it. SG.
(1) -- Would it make sense to check with other "preprocessing authors" (here or in tomorrow's meeting)?

**Comment by ajreynol** [comment]
Yes, this PR indeed needs some discussion at the meeting.

File: src/theory/bv/theory_bv.cpp
Comment (aniemetz) [reply src/theory/bv/theory_bv.cpp:203]: So this only affects preprocessing passes then, right?

File: src/theory/datatypes/theory_datatypes.cpp
Comment (mpreiner) [inline src/theory/datatypes/theory_datatypes.cpp:435]: Looks like this is not formatted.

File: src/theory/bv/theory_bv.cpp
Comment (ajreynol) [reply src/theory/bv/theory_bv.cpp:203]: It impacts preprocessing and `get-value`/`get-model`.

**Review by nafur — APPROVED** [review]
LGTM. Have you run benchmarks?

**Comment by ajreynol** [comment]
I haven't run benchmarks yet.

**Review by yoni206 — APPROVED** [review]
In the meeting we decided it is ok to merge after SMT-LIB run with and without debug assertions. 
LGTM, but lets keep the `do not merge` flag until after runs are done.

**Comment by nafur** [comment]
Current results (with 5s timeout) for non-incremental: [results.txt](https://github.com/CVC4/CVC4/files/5653546/results.txt)

**Comment by ajreynol** [comment]
Based on @nafur 's results, there are no wrong answers or errors, no serious changes in performance (run with a 5 second timeout, the status changes are in QF_BV but everything is close to timeout).  This should be ready to merge.

**Comment by yoni206** [comment]
Thanks, @nafur for all the testing for this!

**Pull request merged**
